Description: Allow choosing and restoring the main menu button icon
Author: Spaced Linux Developers <dev@spacedlinux.com>
Bug: https://github.com/crhy/spaced/issues/379
Forwarded: no

--- a/data/com.solus-project.brisk-menu.gschema.xml
+++ b/data/com.solus-project.brisk-menu.gschema.xml
@@ -46,6 +46,11 @@
       <summary>Keyboard shortcut</summary>
       <description>Accelerator key for opening and closing the menu.</description>
     </key>
+    <key type="s" name="menu-icon">
+      <default>'start-here-symbolic'</default>
+      <summary>Menu button icon</summary>
+      <description>Icon theme name or absolute path to an image file.</description>
+    </key>
     <key type="s" name="label-text">
       <default>""</default>
       <summary>Button label text</summary>
--- a/src/mate-applet/applet.h
+++ b/src/mate-applet/applet.h
@@ -60,6 +60,8 @@
 GType brisk_menu_applet_get_type(void);
 
 void brisk_menu_applet_edit_menus(GtkAction *action, BriskMenuApplet *applet);
+void brisk_menu_applet_reset_icon(GtkAction *action, BriskMenuApplet *applet);
+void brisk_menu_applet_change_icon(GtkAction *action, BriskMenuApplet *applet);
 void brisk_menu_applet_show_about(GtkAction *action, BriskMenuApplet *applet);
 
 G_END_DECLS
--- a/src/mate-applet/main.c
+++ b/src/mate-applet/main.c
@@ -55,6 +55,22 @@
                                                     G_CALLBACK(brisk_menu_applet_edit_menus),
                                                 },
                                                 {
+                                                    "ResetMenuIcon",
+                                                    GTK_STOCK_REVERT_TO_SAVED,
+                                                    N_("Restore Default Menu Icon"),
+                                                    NULL,
+                                                    NULL,
+                                                    G_CALLBACK(brisk_menu_applet_reset_icon),
+                                                },
+                                                {
+                                                    "ChangeMenuIcon",
+                                                    GTK_STOCK_EDIT,
+                                                    N_("Change Menu _Icon"),
+                                                    NULL,
+                                                    NULL,
+                                                    G_CALLBACK(brisk_menu_applet_change_icon),
+                                                },
+                                                {
                                                     "ShowAbout",
                                                     GTK_STOCK_ABOUT,
                                                     N_("_About"),
@@ -71,6 +87,8 @@
  */
 #define BRISK_MENU_XML                                                                             \
         "<menuitem name=\"Edit Menus\" action=\"EditMenus\" />"                                    \
+        "<menuitem name=\"Change Menu Icon\" action=\"ChangeMenuIcon\" />" \
+        "<menuitem name=\"Restore Default Menu Icon\" action=\"ResetMenuIcon\" />" \
         "<menuitem name=\"About\" action=\"ShowAbout\" />"
 
 static gboolean brisk_menu_applet_factory(MatePanelApplet *applet, const gchar *id,
--- a/src/mate-applet/applet.c
+++ b/src/mate-applet/applet.c
@@ -104,6 +104,11 @@
                          self);
 
         g_signal_connect(self->settings,
+                         "changed::menu-icon",
+                         G_CALLBACK(brisk_menu_applet_settings_changed),
+                         self);
+
+        g_signal_connect(self->settings,
                          "changed::window-type",
                          G_CALLBACK(brisk_menu_applet_settings_changed),
                          self);
@@ -139,6 +144,7 @@
         /* Image appears first always */
         image = gtk_image_new_from_icon_name("start-here-symbolic", GTK_ICON_SIZE_MENU);
         self->image = image;
+        brisk_menu_applet_settings_changed(self->settings, "menu-icon", self);
         gtk_box_pack_start(GTK_BOX(layout), image, FALSE, FALSE, 0);
         gtk_widget_set_margin_end(image, 4);
         gtk_widget_set_halign(image, GTK_ALIGN_START);
@@ -255,6 +261,17 @@
                 } else {
                         gtk_label_set_text(GTK_LABEL(self->label), value);
                 }
+        } else if (g_str_equal(key, "menu-icon")) {
+                gchar *value = g_settings_get_string(settings, key);
+                GIcon *icon = g_icon_new_for_string(value, NULL);
+                if (icon) {
+                        gtk_image_set_from_gicon(GTK_IMAGE(self->image), icon, GTK_ICON_SIZE_MENU);
+                        g_object_unref(icon);
+                } else {
+                        gtk_image_set_from_icon_name(GTK_IMAGE(self->image),
+                                                     "start-here-symbolic", GTK_ICON_SIZE_MENU);
+                }
+                g_free(value);
         } else if (g_str_equal(key, "window-type")) {
                 gtk_widget_hide(self->menu);
                 g_clear_pointer(&self->menu, gtk_widget_destroy);
@@ -306,6 +323,47 @@
         }
 
         gtk_image_set_pixel_size(GTK_IMAGE(self->image), final_size);
+}
+
+void brisk_menu_applet_reset_icon(__brisk_unused__ GtkAction *action, BriskMenuApplet *self)
+{
+        g_settings_reset(self->settings, "menu-icon");
+}
+
+/* Store the user's file separately from the theme's start-here icon. */
+void brisk_menu_applet_change_icon(__brisk_unused__ GtkAction *action, BriskMenuApplet *self)
+{
+        GtkWidget *dialog;
+        GtkFileFilter *filter;
+        gint response;
+
+        g_object_ref(self);
+        dialog = gtk_file_chooser_dialog_new(_("Choose Menu Icon"), NULL,
+                                             GTK_FILE_CHOOSER_ACTION_OPEN,
+                                             _("_Cancel"), GTK_RESPONSE_CANCEL,
+                                             _("_Open"), GTK_RESPONSE_ACCEPT, NULL);
+        filter = gtk_file_filter_new();
+        gtk_file_filter_set_name(filter, _("Images"));
+        gtk_file_filter_add_pixbuf_formats(filter);
+        gtk_file_chooser_add_filter(GTK_FILE_CHOOSER(dialog), filter);
+        response = gtk_dialog_run(GTK_DIALOG(dialog));
+        if (response == GTK_RESPONSE_ACCEPT) {
+                gchar *filename = gtk_file_chooser_get_filename(GTK_FILE_CHOOSER(dialog));
+                GError *error = NULL;
+                GdkPixbuf *image = filename ? gdk_pixbuf_new_from_file_at_scale(filename, 256, 256,
+                                                                              TRUE, &error) : NULL;
+                if (image) {
+                        g_settings_set_string(self->settings, "menu-icon", filename);
+                        g_object_unref(image);
+                } else {
+                        brisk_menu_applet_notify_fail(_("Cannot load menu icon"),
+                                                     error ? error->message : _("Select an image file"));
+                }
+                g_clear_error(&error);
+                g_free(filename);
+        }
+        gtk_widget_destroy(dialog);
+        g_object_unref(self);
 }
 
 void brisk_menu_applet_edit_menus(__brisk_unused__ GtkAction *action, BriskMenuApplet *self)
